### DISCUSSION TITLE
Running twister locally

### DISCUSSION THREAD

**Discussion by jachatzi** [zephyrproject-rtos/zephyr#45368] [Q&A]
Hello All,

I'm currently working on my very first PR here. 
The contribution guideline says one shall run twister locally before starting the PR by running:
```
source zephyr-env.sh
./scripts/twister
```

So I did yesterday evening and soon saw first fails, which however seemed not to be related to my changes. Hence I started another machine and started the twister tests there too, but this time with the state of master of which i branched off.
Soon I saw falling tests there too. Also the tests took ages on both machines. So i called it a day.

Fast forward to this morning: I found both machines with full disks (both machines are rather potent in regards of processing power and disk size.)

Thus, now I'm wondering: Have I done something wrong or does the instruction in the contribution guidelines come from older times where there weren't that much tests and one could easily run them all locally?

**Comment by yashi** [comment]
You can run a test you are interested in by using the option `-T`, with your architecture `-a`.  You can also list all tests using `--list-tests`.  Run `twister -h` to see all options. Also, `-v` tells you what's going on underneath.

This is the line I've used to test kernel tests against ARC arch:

```
zephyr/scripts/twister -vvvv -N -j1 -a arc -T zephyr/tests/kernel/common/
```

Depending on the tests and arch, `twister-out.*` directories will hold a quite bit of data.  You can remove them if you don't need .

**Comment by jachatzi** [comment]
Thanks a lot for the wrap up of how to restrict the tests to the tests of interest, I'll definitely use that in the future.

I'm asking myself, however, is if the contribution guideline should be changed to state that one should "identify the tests which could be affected by ones change and run those", instead of giving the commands to run all tests and thus send a newbie down a road of full disks :)

**Comment by AndreyDodonov-EH** [comment]
@jachatzi 
I think one can define which tests to run using `/scripts/ci/test_plan.py`  script.
That's how PR defines which tests to run.
E.g. I opened a PR for TCP, and then this script was called.

https://github.com/zephyrproject-rtos/zephyr/runs/6335112164?check_suite_focus=true#step:12:5
